cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
     }
 
     Set<String> remainingSources = Sets.intersection(remainingNodes, possibleNewSources);
+    Set<String> processedNodes = new HashSet<>();
+    // Since there can be remaining sources from subdags which don't overlap, they should be split as seperate subdags.
+    // For that, we first create all subdags from remaining sources and keep track of accessible nodes for those
+    // subdags. Then we merge overlapping subdags
     if (!remainingSources.isEmpty()) {
-      dags.add(subsetFrom(remainingSources, possibleNewSinks));
+      // source -> [ nodes accessible by the source ]
+      Map<String, Set<String>> nodesAccessibleBySources = new HashMap<>();
+
+      for (String remainingSource : remainingSources) {
+        Dag remainingNodesDag = subsetFrom(ImmutableSet.of(remainingSource), possibleNewSinks);
+        nodesAccessibleBySources.put(remainingSource, accessibleFrom(remainingSource, remainingNodesDag.getSinks()));
+      }
+
+      for (String remainingSource : remainingSources) {
+        if (!processedNodes.contains(remainingSource)) {
+          Set<String> remainingAccessibleNodes = nodesAccessibleBySources.get(remainingSource);
+          // Island of all the remaining nodes
+          Set<String> islandNodes = new HashSet<>();
+          islandNodes.addAll(remainingAccessibleNodes);
+          for (String otherSource : remainingSources) {
+            if (!remainingSource.equalsIgnoreCase(otherSource)) {
```

Comment:
nit: a little easier to read if you do

```
if (remainingSource.equals(otherSource)) {
  continue;
}
```

since it removes one layer of nesting